core/app/jobs/spree/promotions/duplicate_promotion_job.rb

```diff
@@ -0,0 +1,49 @@
+module Spree
+  module Promotions
+    class DuplicatePromotionJob < Spree::BaseJob
+      def perform(template_promotion_id:, batch_id:, options: {}, code: nil)
+        promotion = find_instance(template_promotion_id, model: Spree::Promotion)
+        code = code || applicable_code(batch_id, options)
+
+        Spree::PromotionHandler::PromotionBatchDuplicator.new(promotion, batch_id, code: code).duplicate
+      rescue CodeGenerator::MutuallyExclusiveInputsError, ActiveRecord::RecordNotUnique => e
+        Rails.logger.error(e.message)
+      end
+
+      private
+
+      def find_instance(id, model:)
```

Comment:
Common interface for finding either a `Promotion` or a `PromotionBatch`, so that the code adheres to the good practice of DRY.
Also, I've concluded these calls should be extracted from the `perform` so that the `perform` will have less responsibilities (so S from SOLID).